core/dbt/deps/local.py

```diff
@@ -39,9 +41,11 @@ def resolve_path(self, project):
             project.project_root,
         )
 
-    def _fetch_metadata(self, project, renderer):
-        loaded = project.from_project_root(self.resolve_path(project), renderer)
-        return ProjectPackageMetadata.from_project(loaded)
+    def _fetch_metadata(
```

Comment:
Seems like this function and the function in `git.py` is mostly similar other than the way path is extracted, maybe consider refactor both into one function?

Reply:
I was itching to do the same but I keep going back on forth on whether it makes sense refactored elsewhere. At this point the only deviation between the two _fetch_metadata implementations is in how they construct a `PartialProject` to call `render_package_metadata` on. Figuring out how the `PartialProject` gets constructed (if one is necessary at all) feels like the right level of responsibility for the `PinnedPackage` subclasses.

Additionally, I'm not sure where a better fit for this common logic would be. A couple alternatives considered: 
1. In `ProjectPackageMetadata`, perhaps something like `ProjectPackageMetadata.from_project_dir(project_dir, renderer)`. This is what I went with initially but ran into circular imports between `dbt.config` and `dbt.contracts`. The logic also felt better contained in `PartialProject` where the other relevant helpers were already defined. 
2. In the common base class, something like `PinnedPackage._fetch_metadata_from_project_dir(project_dir, renderer)`. It doesn't feel great for the base class to be implemented with so much awareness of what just a subset of its subclasses are doing.

Reply:
I was thinking of the second option you have, didn't realize that there's also a `registry` method also overwrite this function. Although it looks like that one doesn't even use the two arguments passed in.
I agree that this doesn't feel like something we can come up with a easy-to-implement better solution very soon. So I am okay leaving it as is for now.